gms/gossiper.cc

```diff
@@ -737,36 +736,45 @@ void gossiper::run() {
 
                 _gossiped_to_seed = false;
 
-                auto get_random_node = [this] (const utils::chunked_vector<inet_address>& nodes) {
-                    std::uniform_int_distribution<int> dist(0, nodes.size() - 1);
-                    int index = dist(this->_random_engine);
-                    return nodes[index];
-                };
-
-                /* Gossip to some random live members */
-                // TODO: For now, we choose 10th of all the nodes in the cluster.
-                auto nr_live_nodes = std::max(size_t(1), endpoint_state_map.size() / 10);
-                nr_live_nodes = std::min(nr_live_nodes, _live_endpoints.size());
-                std::unordered_set<gms::inet_address> live_nodes;
-                logger.debug("nr_live_nodes={}, endpoint_state_map.size()={}, live_endpoints.size={}",
-                    nr_live_nodes, endpoint_state_map.size(), _live_endpoints.size());
-                while (live_nodes.size() < nr_live_nodes && nr_live_nodes <= _live_endpoints.size()) {
-                    if (!_live_endpoints_just_added.empty()) {
-                        auto ep = _live_endpoints_just_added.front();
-                        _live_endpoints_just_added.pop_front();
-                        logger.debug("Favor newly added node {}", ep);
-                        live_nodes.insert(ep);
-                    } else {
-                        // Get a random live node
-                        live_nodes.insert(get_random_node(_live_endpoints));
+                if (_endpoints_to_talk_with.empty()) {
+                    std::shuffle(_live_endpoints.begin(), _live_endpoints.end(), _random_engine);
+                    // This guarantees the local node will talk with all nodes in live_endpoints once within nr_rounds gossip rounds.
+                    size_t nr_rounds = 10;
```

Comment:
I asked for a link to material you based this on. Both for the review, and as in-code explanation of what is going on.

Reply:
OK. I thought you just wanted the link in the PR comments.